### ISSUE TITLE
Support dataset creation using `cuCIM`

### ISSUE THREAD

**Issue by Leengit** [DigitalSlideArchive/HistomicsStream#74]
We want to add `cuCIM` functionality as an alternative to use of `large_image` to create a `TensorFlow` `Dataset`.

# Background:
Current `HistomicsStream` functionality can be partitioned into two parts 

1. Specify the configuration (which slides to use, tile size, amount of overlap, mask, ...), which has everything but the actual pixel data from the whole-slide images.
2. Take a fully-specified configuration and turn it into a `TensorFlow` `Dataset` that lazily provides access to both the actual pixel data and the configuration information.  (It is "lazy" in the sense that we will be using the graph mode of `TensorFlow` rather than its eager mode.)

Much useful information about the current usage of `HistomicsStream` can be found at [`README.md`](https://github.com/DigitalSlideArchive/HistomicsStream/blob/master/README.md) and [`StudyObject`](https://github.com/DigitalSlideArchive/HistomicsStream/blob/master/StudyObject.md).

# Recommended Approach
In this case, we want as an alternative second step to use `cuCIM` to create a `TensorFlow` `Dataset` that has lazy access to the pixel data.  Alternatively, if that proves difficult we could instead create a dataset for `PyTorch` or `MONAI` as a first task.  Ultimately, assuming `cuCIM` continues to be promising, we will want to be able support all three of these platforms.

The current [`tensorflow_stream.ipynb`](https://github.com/DigitalSlideArchive/HistomicsStream/blob/master/example/tensorflow_stream.ipynb) file should be duplicated and checked in as a new file called something like `cucim_stream.ipynb`.  The part of the notebook that is 
```python
create_tensorflow_dataset = hs.tensorflow.CreateTensorFlowDataset()
tiles = create_tensorflow_dataset(my_study_of_tiles)
```
should be discarded and replaced with your new code that instead uses `cuCIM`.  That is, you should put the `cuCIM` code in a separate `histomics_stream/cucim.py` file that is imported to the notebook, and the notebook itself will use a few lines of code to invoke the code from that new file.

At any point where you would like someone to look over your code, please push your new git branch to your GitHub fork of the [git@github.com:DigitalSlideArchive/HistomicsStream.git](https://github.com/DigitalSlideArchive/HistomicsStream.git) repository.  Then create a draft pull request.

# More discussion
This Issue is a good place for clarifying the goals and approach of this task.  Simply reply here.

**Comment by Leengit** [comment]
@cooperlab @thewtex please feel free to add any details or corrections.

**Comment by cooperlab** [comment]
Thank you @Leengit for the nice summary.

We essentially want to figure out what parameters to use when calling [CuImage.read_region](https://github.com/rapidsai/cucim/blob/4061b7e00c5b10f18f04f3c356b32a42c081de4f/cpp/include/cucim/cuimage.h#L179). Step 1 would be to understand the configuration object that HistomicsStream generates, and how the parameters are fed to large_image to get the correct regions. From there we should be able to map these parameters to the correct ones for CuImage.

I think the CuImage cache may persist only as long as the CuImage object, so if we are creating a new CuImage object to generate each tile we might not see the benefits of this. Large_image has a similar cache. How does this currently work?

**Comment by Leengit** [comment]
> I think the CuImage cache may persist only as long as the CuImage object, so if we are creating a new CuImage object to generate each tile we might not see the benefits of this. Large_image has a similar cache. How does this currently work?

[At Lines 57-58](https://github.com/DigitalSlideArchive/HistomicsStream/blob/cc34b14ac6f8c4c0effdb829636932c09b7914d6/histomics_stream/tensorflow.py#L57-L58), histomics_stream modifies the configuration description for the purpose of partitioning the tiles based upon the  chunk that each will belong to.  We may want to have more-or-less the exact step for the cucim pipeline.  (Aside: If the two steps are exactly the same, perhaps break out the `_designate_chunks_for_tiles functionality` into a new class and into a new file other than `tensorflow.py` to indicate that its purpose is more general than the `tensorflow.py` file's purpose.  Perhaps the new class will become the superclass to the main classes of both `tensorflow.py` and `cucim.py` so that it is readily available to both implementations, and yet the writer of a Jupyter notebook doesn't have to know anything about it..)

[At Lines 62-113](https://github.com/DigitalSlideArchive/HistomicsStream/blob/cc34b14ac6f8c4c0effdb829636932c09b7914d6/histomics_stream/tensorflow.py#L62-L113) a `TensorFlow` `Dataset` is created with one record per chunk.  This does not have the pixel by pixel data yet but it is now a genuine `TensorFlow` `Dataset`.  I don't know the cuCIM paradigm well enough; but we may want to implement a very similar step using cuCIM functions.

The next step is use `map` functionality of `TensorFlow` `Dataset` to lazily read in the pixel data, one chunk at a time -- see [Lines 330](https://github.com/DigitalSlideArchive/HistomicsStream/blob/cc34b14ac6f8c4c0effdb829636932c09b7914d6/histomics_stream/tensorflow.py#L330) and following.  I don't know the cuCIM paradigm well enough; but we may want to implement a very similar step, instead using cuCIM functions.  (As mentioned above, by "lazily read" the idea is to not actually read in the pixel data at this time, but to construct the TensorFlow execution graph with the commands that will instead do the reading at that time that TensorFlow determines is the optimal time to do that.  I don't know the extent to which that changes with cuCIM.)

The next step is  to lazily split each chunk's record (including its pixel data) into one record for each tile -- see [Line 243](https://github.com/DigitalSlideArchive/HistomicsStream/blob/cc34b14ac6f8c4c0effdb829636932c09b7914d6/histomics_stream/tensorflow.py#L243) and following.  In `tensorflow.py`, this is immediately followed by an `unbatch` command that combines each chunk's list of tiles into one giant list of all tiles.  Each tile record is then rearranged so that typical `TensorFlow` models will find the parts they need in standard locations.  Because the steps in this paragraph do not involve reading in the pixel data from its source, the steps may be close to identical in the current and new implementations (and thus might end up being part of a superclass's functionality).

**Comment by cooperlab** [comment]
@Leengit Thank you. I think we currently create a new LargeImage object for each chunk that is read. We did the same for OpenSlide, and would do the same for CuImage (the interface is very similar). I don't see any way around creating a CuImage or LargeImage object for each chunk, and caching would only be beneficial when reading multiple chunks. It may not matter at all - if chunks are large then there are fewer seams where the caching can help anyway.

**Comment by Leengit** [comment]
The part of this task that I don't understand how to solve is how to tell `TensorFlow` that the pixel data that we have used `cuCIM` to load into GPU memory comprise a TensorFlow `tensor`.  That is, I don't see any constructor (or any similar function) for `tensor` or `Dataset` that allows one to provide a pointer into GPU memory.  I don't see any examples on the `cuCIM` web site that indicate how to do this.  Hopefully, I have merely overlooked them.  Or, maybe, NVIDIA `DALI` provides the needed functionality; I haven't looked to any depth there.

**Comment by cooperlab** [comment]
Let's make sure the buffer is on device before we go down this path. The device argument has to be passed value "cuda" for nvJPEG GPU decoding.

**Comment by thewtex** [comment]
To investigate: cuCIM / Tensorflow transfer via the [__cuda_array_interface](https://numba.pydata.org/numba-doc/dev/cuda/cuda_array_interface.html).

**Comment by Leengit** [comment]
The cuCIM folks also point us towards DLPack.  They indicate that there is support in TensorFlow via the `tfdlpack-gpu` package.  Hopefully this addresses the "telling TensorFlow about the tensors already on the GPU" issue that we have been discussing.

**Comment by Leengit** [comment]
@cooperlab finds:
> There is a comment in tf-dlpack that we should keep in mind - "Currently only tested under tensorflow 2.0's eager mode". We often run HistomicsStream in graph-mode.

The ability of HistomicsStream to handle large images is significantly lessened in eager mode ... if we can't get tfdlpack-gpu to work in TensorFlow graph mode then that is a big negative.

**Comment by Leengit** [comment]
Even if dlpack is seemingly only for TensorFlow eager mode we may be able to use it in graph mode by invoking the dlpack functionality within a function that is called via `tf.py_function` from a function annotated as `@tf.function`.  For example, see https://github.com/DigitalSlideArchive/HistomicsStream/blob/cc34b14ac6f8c4c0effdb829636932c09b7914d6/histomics_stream/tensorflow.py#L221-L242

**Comment by diricxbart** [comment]
> Even if dlpack is seemingly only for TensorFlow eager mode we may be able to use it in graph mode by invoking the dlpack functionality within a function that is called via `tf.py_function` from a function annotated as `@tf.function`.

Please see some consequences of the suggested approach in this comment: https://github.com/tensorflow/tensorflow/issues/51642#issuecomment-922965469

> Note however that py_function code is not portable, and I don't think it will work well with tf.distribute. For use with tf.function / tf.distribute, my advice would be to rewrite the CuPy code into TF ops. Fundamentally, they should be similar - in both cases you'd run GPU kernels.

Nevertheless, it might still be the best way forward...

**Comment by ahmadrathore** [comment]
Running the example code of converting a Tensorflow tensor to Torch tensor using DLPack from https://github.com/VoVAllen/tf-dlpack, RuntimeError of 'Unsupported tensorflow version:' comes at the line: 
import tfdlpack.
It happens both on the Docker notebook and CoLab. I have tried latest and old version of tensorflow.
Is there a way around for this?

**Comment by Leengit** [comment]
It looks like tfdlpack is no longer an external package.  Perhaps https://www.tensorflow.org/api_docs/python/tf/experimental/dlpack or a related page shows how to use it after a mere
```python
import tensorflow as tf
```

**Comment by cooperlab** [comment]
> 



> 

Thank you @diricxbart for pointing this out. We wrap other Python libraries in our TensorFlow pipeline and so it is likely already not portable.

**Comment by Leengit** [comment]
@cooperlab says to skip this because large_image > cuCIM.
